source/FreeRTOS_TCP_IP.c

```diff
@@ -272,39 +272,56 @@
     void vTCPStateChange( FreeRTOS_Socket_t * pxSocket,
                           enum eTCP_STATE eTCPState )
     {
-        FreeRTOS_Socket_t * xParent = NULL;
+        FreeRTOS_Socket_t * xParent = pxSocket;
         BaseType_t bBefore = tcpNOW_CONNECTED( ( BaseType_t ) pxSocket->u.xTCP.eTCPState ); /* Was it connected ? */
         BaseType_t bAfter = tcpNOW_CONNECTED( ( BaseType_t ) eTCPState );                   /* Is it connected now ? */
 
-        #if ( ipconfigHAS_DEBUG_PRINTF != 0 )
-            BaseType_t xPreviousState = ( BaseType_t ) pxSocket->u.xTCP.eTCPState;
-        #endif
+        BaseType_t xPreviousState = ( BaseType_t ) pxSocket->u.xTCP.eTCPState;
+
         #if ( ipconfigUSE_CALLBACKS == 1 )
             FreeRTOS_Socket_t * xConnected = NULL;
         #endif
 
+        if( ( ( xPreviousState == eCONNECT_SYN ) ||
+              ( xPreviousState == eSYN_FIRST ) ||
+              ( xPreviousState == eSYN_RECEIVED ) ) &&
+            ( eTCPState == eCLOSE_WAIT ) )
+        {
+            /* A socket was in the connecting phase but something
+             * went wrong and it should be closed. */
+            FreeRTOS_printf( ( "Move from %s to %s\n",
+                               FreeRTOS_GetTCPStateName( xPreviousState ),
+                               FreeRTOS_GetTCPStateName( eTCPState ) ) );
+            bBefore = pdTRUE;
```

Comment:
```suggestion
            /* Set the flag to show that it was connected before and that the
             * status has changed now. This will cause the control flow to go 
             * in the below if condition.*/
            bBefore = pdTRUE;
```